File: src/handlers/drawHandler.js
import { spawn } from 'child_process';
import { fileURLToPath } from 'url';
import { dirname, join, resolve } from 'path';
import { existsSync, mkdirSync } from 'fs';
import { readFile, writeFile } from 'fs/promises';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const projectRoot = resolve(__dirname, '../..');
const drawDir = join(projectRoot, 'draw');
const venvPython = join(drawDir, 'venv', 'bin', 'python3');

/**
 * 调用 Python 脚本绘制所有 topScoreLossMoves 的 GIF
 *
 * @param {string} jsonFilePath - JSON 文件路径（包含 moves 数据）
 * @param {string} outputDir - 输出目录
 * @returns {Promise<Array<string>>} 生成的 GIF 文件路径数组
 */
export async function drawAllMovesGif(jsonFilePath, outputDir = null) {
  // 默认输出目录
  if (!outputDir) {
    outputDir = join(projectRoot, 'draw', 'output');
  }

  // 确保输出目录存在
  if (!existsSync(outputDir)) {
    mkdirSync(outputDir, { recursive: true });
  }

  // 提取并输出文件夹名称
  const dirName = outputDir.split('/').pop() || outputDir.split('\\').pop();
  console.log(`Drawing all moves GIFs to outputDir: ${dirName}`);

  // Python 脚本路径
  const pythonScript = join(drawDir, 'draw.py');

  if (!existsSync(pythonScript)) {
    throw new Error(`Python script not found: ${pythonScript}`);
  }

  if (!existsSync(jsonFilePath)) {
    throw new Error(`JSON file not found: ${jsonFilePath}`);
  }

  // 检查虚拟环境是否存在
  if (!existsSync(venvPython)) {
    throw new Error(
      `Virtual environment not found: ${venvPython}\nPlease create it with: python3 -m venv ${join(
        drawDir,
        'venv'
      )} && ${venvPython} -m pip install -r ${join(
        drawDir,
        'requirements.txt'
      )}`
    );
  }

  return new Promise((resolve, reject) => {
    // 使用虚拟环境中的 Python 解释器调用脚本
    const python = spawn(venvPython, [pythonScript, jsonFilePath, outputDir], {
      cwd: projectRoot,
      stdio: ['inherit', 'pipe', 'pipe']
    });

    let stdout = '';
    let stderr = '';

    python.stdout.on('data', (data) => {
      const output = data.toString();
      stdout += output;
      // 实时输出到 shell，只显示文件名
      const lines = output.split('\n');
      lines.forEach((line) => {
        const match = line.match(/GIF created: (.+)/);
        if (match) {
          const fullPath = match[1];
          const filename =
            fullPath.split('/').pop() || fullPath.split('\\').pop();
          process.stdout.write(`GIF created: ${filename}\n`);
        } else if (line.trim()) {
          // 其他输出原样显示
          process.stdout.write(line + '\n');
        }
      });
    });

    python.stderr.on('data', (data) => {
      const output = data.toString();
      stderr += output;
      // 实时输出到 shell
      process.stderr.write(output);
    });

    python.on('close', (code) => {
      if (code === 0) {
        // 从 stdout 中提取所有生成的 GIF 路径
        const gifMatches = stdout.matchAll(/GIF created: (.+)/g);
        const gifPaths = Array.from(gifMatches, (match) => match[1]);
        resolve(gifPaths);
      } else {
        reject(new Error(`Python script failed with code ${code}\n${stderr}`));
      }
    });

    python.on('error', (error) => {
      reject(new Error(`Failed to start Python script: ${error.message}`));
    });
  });
}

File: apps/localhost_all/src/handlers/draw_handler.py
import os
import sys
import asyncio
import re
import subprocess
from pathlib import Path
from typing import List


async def draw_all_moves_gif(json_file_path: str, output_dir: str = None) -> List[str]:
    """Call Python script to draw GIFs for all topScoreLossMoves"""
    # Get project root
    current_file = Path(__file__)
    project_root = current_file.parent.parent.parent
    
    # Default output directory
    if not output_dir:
        output_dir = str(project_root / "draw" / "output")
    
    # Ensure output directory exists
    os.makedirs(output_dir, exist_ok=True)
    
    # Extract and output folder name
    dir_name = os.path.basename(output_dir)
    print(f"Drawing all moves GIFs to outputDir: {dir_name}")
    
    # Python script path
    draw_dir = project_root / "draw"
    python_script = draw_dir / "draw.py"
    
    if not python_script.exists():
        raise FileNotFoundError(f"Python script not found: {python_script}")
    
    if not os.path.exists(json_file_path):
        raise FileNotFoundError(f"JSON file not found: {json_file_path}")
    
    # Use the current Python interpreter (same as the running process)
    python_executable = sys.executable
    
    # Use Python interpreter to call script
    process = await asyncio.create_subprocess_exec(
        python_executable,
        str(python_script),
        json_file_path,
        output_dir,
        cwd=str(project_root),
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE
    )
    
    stdout = b""
    stderr = b""
    
    # Capture stdout
    while True:
        chunk = await process.stdout.read(1024)
        if not chunk:
            break
        stdout += chunk
        output = chunk.decode("utf-8", errors="replace")
        # Real-time output to shell, only show filename
        lines = output.split("\n")
        for line in lines:
            match = re.search(r"GIF created: (.+)", line)
            if match:
                full_path = match.group(1)
                filename = os.path.basename(full_path)
                print(f"GIF created: {filename}")
            elif line.strip():
                # Other output as-is
                print(line)
    
    # Capture stderr
    while True:
        chunk = await process.stderr.read(1024)
        if not chunk:
            break
        stderr += chunk
        # Real-time output to shell
        print(chunk.decode("utf-8", errors="replace"), end="", file=sys.stderr)
    
    return_code = await process.wait()
    
    if return_code == 0:
        # Extract all generated GIF paths from stdout
        stdout_text = stdout.decode("utf-8", errors="replace")
        gif_matches = re.findall(r"GIF created: (.+)", stdout_text)
        return gif_matches
    else:
        raise RuntimeError(
            f"Python script failed with code {return_code}\n{stderr.decode('utf-8', errors='replace')}"
        )

